okhttp/src/main/java/okhttp3/EventListener.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3;
+
+import java.net.InetAddress;
+import java.util.List;
+
+/**
+ * Observes, modifies, and potentially short-circuits requests going out and the corresponding
+ * responses coming back in. Typically interceptors add, remove, or transform headers on the request
+ * or response.
+ */
+public class EventListener {
```

Comment:
maybe add javadoc for each callback, mentioning whether they are re-entrant or not and if they are potentially terminal.

For example, in the interceptors design, the application interceptor is called once per semantic call, and the network interceptor is called once (or maybe not when served from cache?)

This will help us understand the state model. For example, if we do a single-span for the operation, will fetchEnd always be called, even if dnsEnd has a throwable?